File: 0733-flood-fill/0733-flood-fill.py
from collections import deque

class Solution:

    def dfs(self, image: List[List[int]], sr: int, sc: int, origin: int, color: int) -> None:
        dx = [-1,0,1,0]
        dy = [0,-1,0,1]

        n = len(image)
        m = len(image[0])

        dq = deque([])
        dq.append((sr,sc))
        visited = [[False] * m for _ in range(n)] # 방문 여부 체크 이차원 배열
        visited[sr][sc] = True

        while dq:
            cur_x, cur_y = dq.popleft()
            image[cur_x][cur_y] = color

            for i in range(4):
                nx = dx[i] + cur_x
                ny = dy[i] + cur_y

                if nx < 0 or nx >= n or ny < 0 or ny >= m: # 범위를 벗어나는 경우
                    continue
                if visited[nx][ny]: # 이미 방문한 경우
                    continue
                if image[nx][ny] != origin: continue # 출발점과 색이 다른 경우
                
                visited[nx][ny] = True
                dq.append((nx,ny))






    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        self.dfs(image, sr, sc, image[sr][sc], color)
        return image
        

File: 0733-flood-fill/0733-flood-fill.java
class Solution {

    int[] dx = {-1,0,1,0};
    int[] dy = {0,-1,0,1};

    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        bfs(image, new Point(sr,sc), image[sr][sc], color);

        return image;
    }

    public void bfs(int[][] image, Point start, int origin, int color){
        int n = image.length;
        int m = image[0].length;
        boolean[][] visited = new boolean[n][m];

        Queue<Point> q = new LinkedList<>();
        q.add(start);
        visited[start.x][start.y] = true;

        while(!q.isEmpty()){
            Point cur = q.poll();
            image[cur.x][cur.y] = color;

            for(int i = 0; i < 4; i++){
                int nx = dx[i] + cur.x;
                int ny = dy[i] + cur.y;
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue; // 범위를 벗어난 경우
                if(visited[nx][ny]) continue; // 이미 방문한 경우
                if(image[nx][ny] != origin) continue;
                visited[nx][ny] = true;
                q.add(new Point(nx,ny));
            }
        }
    }
}

class Point{
    int x,y;

    public Point(int x, int y){
        this.x = x;
        this.y = y;
    }
}